Modules.md

```diff
@@ -0,0 +1,160 @@
+# Modules
+
+## Summary
+
+We propose adding a module system mechanism to the WGSL shading language as an extension.
+
+## Assumptions
+
+Assumes that [`load`](./Imports.md) has already been implemented.
+
+# Motivation
+
+Currently all symbols in wgsl share a single global namespace with a prohibition against symbols with the same name. Many existing implementations of imports (not unlike [our proposal](./Imports.md)) simply add to this global scope. For relatively self contained projects this is not a problem, however when considering a broader ecosystem of packages containing reusable shaders, collisions become much more likely. 
+
+Additionally WGSL provides very little way to encapsulate and organise code. This proposal would pave the way for both
+[Module Interfaces](./ModulesInterfaces.md) and [Generic Modules](./GenericModules.md). The former would allow a graphics programmer to optionally control the visibility and typecheck the symbols within a module, while the latter would allow for more reusable code and the ability to build powerful abstractions. 
```

Comment:
It's impressive how you show the mod scaffolding can be stretched to cover so many features. 
What's the +/- for building those features as 'mod' extensions vs building them separately?

Reply:
While hard to list the pros and cons without having something concrete to compare it against, there are a few things I'd call out. 

On the plus side, the contents of modules are valid wgsl; there are no additional constructs to transpile other than mangling the symbols within a module. 

On the minus is that sometimes it may be more verbose than e.g. generic functions — though this could be solved by syntactic sugar that coerces functions and constants into modules. 

Another minus is that this style of programming may be unfamiliar to many graphics programmers. Though this is somewhat ameliorated by having few additional concepts in the WESL language other than modules.